File: app/src/main/java/com/pk/dexloadexemplar/DexDisposableObserver.java
package com.pk.dexloadexemplar;

import android.content.Context;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.pk.dexloadexemplar.DLog.DLog;
import com.pk.dexloadexemplar.apk.ApkFile;

import io.reactivex.annotations.NonNull;
import io.reactivex.observers.DisposableObserver;

public class DexDisposableObserver extends DisposableObserver<Double> {
    private static final String TAG = DexDisposableObserver.class.getName();
    ApkFile apkFile;
    ProgressBar progressBar;
    Context context;

    public DexDisposableObserver(
            Context context,
            ApkFile apkFile,
            ProgressBar progressBar
    ) {
        this.apkFile = apkFile;
        this.progressBar = progressBar;
        this.context = context;
    }

    @Override
    public void onNext(@NonNull Double aDouble) {
        DLog.d(TAG, "onNext", "Progress: " + aDouble);
        progressBar.setProgress((int) (aDouble*100));
    }

    @Override
    public void onError(@NonNull Throwable e) {
        DLog.e(TAG, "onError", "Error: " + e);
    }

    @Override
    public void onComplete() {
        DLog.i(TAG, "onComplete", "completed");
        progressBar.setProgress(0);
        progressBar.post(() -> {
            Toast.makeText(context, "Download complete", Toast.LENGTH_SHORT).show();
        });
    }
}

File: app/src/main/java/com/pk/dexloadexemplar/DexDisposableObserver.kt
package com.pk.dexloadexemplar

import android.content.Context
import android.view.View
import android.widget.ProgressBar
import android.widget.Toast
import com.pk.dexloadexemplar.DLog.DLog
import com.pk.dexloadexemplar.apk.ApkFile
import com.pk.dexloadexemplar.extract.ExtractTask
import com.pk.dexloadexemplar.runnerTask.DexLauncherTask
import io.reactivex.observers.DisposableObserver

class DexDisposableObserver(
        var context: Context,
        var apkFile: ApkFile,
        var progressBar: ProgressBar,
        var view: View? = null
) : DisposableObserver<Double?>() {
    override fun onNext(aDouble: Double) {
        DLog.d(TAG, "onNext", "Progress: $aDouble")
        progressBar.progress = (aDouble * 100).toInt()
    }

    override fun onError(e: Throwable) {
        DLog.e(TAG, "onError", "Error: $e")
        progressBar.post {
            view?.isEnabled = true
        }
    }

    override fun onComplete() {
        DLog.i(TAG, "onComplete", "completed")
        progressBar.progress = 0
        progressBar.post {
            Toast.makeText(context, "Download complete", Toast.LENGTH_SHORT).show()
            view?.isEnabled = true
            if (ExtractTask(context).apply(apkFile)) {
                Toast.makeText(context, "Extraction Done", Toast.LENGTH_SHORT).show()
                DexLauncherTask(context).apply(apkFile)
            }
        }

    }

    companion object {
        private val TAG = DexDisposableObserver::class.java.name
    }
}